### ISSUE TITLE
Feature Request: add HTTPS proxy

### ISSUE THREAD

**Issue by s3rj1k** [telegramdesktop/tdesktop#4577]
In addition to HTTP proxy, would be nice to have option to use HTTPS proxy (tls+http)
for more security

**Comment by stek29** [comment]
Same goes for SOCKS (even with passwords) btw.

**Comment by s3rj1k** [comment]
@stek29 yep. But secure SOCKS (shadowsocks) is kind of hackish protocol.

On the other hand HTTPS is defacto standard, implemented in all major web browsers.

**Comment by s3rj1k** [comment]
@john-preston any news on this?

**Comment by john-preston** [comment]
@s3rj1k No news, sorry :( Current proxy implementation works through http://doc.qt.io/qt-5/qnetworkproxy.html and I'm not sure they support https proxy (otherwise it would work already).

**Comment by s3rj1k** [comment]
made JIRA issue for upstream https://bugreports.qt.io/browse/QTBUG-67709

**Comment by stek29** [comment]
@101usb No, it is as safe to use Telegram with proxy as without proxy (so pretty much safe)

**Comment by evgenylarin** [comment]
@stek29 @101usb mean that  proxy credentials could be intercept by MITM and using, because credentials transfers as a plain text.

**Comment by Aokromes** [comment]
@john-preston check if this article can help you https://www.meetingcpp.com/blog/items/http-and-https-in-qt.html

**Comment by john-preston** [comment]
@Aokromes I have HTTP and HTTPS requests used in tdesktop, but I'm not sure how you use HTTPS proxy there - I see only HTTP proxy type in docs.

**Comment by omkarnathsingh** [comment]
I tried a HTTPS proxy from here, https://us-proxy.org/ It worked for me in Tdesktop 1.2.18 alpha, although it displays HTTP in tdesktop UI.

**Comment by s3rj1k** [comment]
I'll just leave this link [here](https://tproger.ru/news/rostelekom-against-mtproto-proxy/) (in russian)

**Comment by lgg** [comment]
@s3rj1k that link' tl;dr;

One of the biggest Russian ISP blocks MTProxy by checking packet sizes.

**Comment by s3rj1k** [comment]
Still an issue

**Comment by ilya-fedin** [comment]
Qt folks decided not to support HTTPS proxies, so there wouldn't be as well:

**Comment by s3rj1k** [comment]
@ilya-fedin Well this is disappointing to say the least.

**Comment by s3rj1k** [comment]
Still an issue, this brings a bit more sadness :(

**Comment by ilya-fedin** [comment]
This still needs implementation on Qt level. Given that Qt folks decided not to implement this, this will be never implemented, most likely.

**Comment by s3rj1k** [comment]
MTProto mode is defiantly not implemented by QT, so what stops using some third-party lib to support HTTPS proxy?

**Comment by ilya-fedin** [comment]
Are you sure there any? According to Qt issue, only browsers have support for this.

**Comment by s3rj1k** [comment]
@ilya-fedin https://github.com/curl/curl/blob/master/docs/cmdline-opts/version.d#L41
curl supports HTTPS proxies

**Comment by ilya-fedin** [comment]
I don't think it's possible to use Qt's networks APIs and curl's proxy support simultaneously. Migrating from Qt's network APIs to curl's one is not an option, most likely.

**Comment by ilya-fedin** [comment]
> Then how is Qt network API currently works with MTProto?

MTProto is implemented from scratch using Qt network APIs. Don't except that the same amount of work will be repeated to support HTTPS proxies, of course.

**Comment by s3rj1k** [comment]
@ilya-fedin No need to implement whole HTTPS support, only bridge between SOCKS5 native support in Qt network APIs and internal SOCKS5 layer to HTTPS, that way any proxy protocol could be implemented without a need of Qt support

**Comment by s3rj1k** [comment]
Would be perfectly fine if telegram desktop would start SOCKS5 proxy bound to localhost

**Comment by ilya-fedin** [comment]
> No need to implement whole HTTPS support, only bridge between SOCKS5 native support in Qt network APIs and internal SOCKS5 layer to HTTPS, that way any proxy protocol could be implemented without a need of Qt support

Well, if you know how to implement, maybe you create a PR?

**Comment by s3rj1k** [comment]
@ilya-fedin If I could, I would have, but sadly I do not have needed expertise in C/C++

I could write one in Go :)

**Comment by ilya-fedin** [comment]
'bridge between SOCKS5 native support in Qt network APIs and internal SOCKS5 layer to HTTPS' doesn't sound like something easy to implement. If HTTPS proxies have SOCKS inside then it's even more complicated than I thought.

**Comment by s3rj1k** [comment]
@ilya-fedin No, I was talking about different thing.

 - TelegramDesktop starts in a separated thread SOCK5 server, that accepts connections on loopback
 - Qt network API calls are pointed to that SOCKS5 server
 - SOCKS5 server acts as a MITM proxy and proxies all cannection to remote HTTPS proxy

That way there is no need to implement a wrapper for Qt network API

**Comment by ilya-fedin** [comment]
I don't think such over-engineering is acceptable

**Comment by s3rj1k** [comment]
@ilya-fedin I will allow to support any feasible network proxy, so I don't think that this is geoengineering.

But yea, less prettier then native Qt API wrapper

**Comment by s3rj1k** [comment]
And Qt already has SOCKS server code bits :)

**Comment by ilya-fedin** [comment]
> so I don't think that this is over-engineering

Having a proxy server inside a messenger is definitely over-engineering. Not to say how much questions would be from admins of stores with confined environment (Microsoft Store, Mac App Store, Snap Store, Flathub).

**Comment by s3rj1k** [comment]
Could be an optional plugin maybe.

**Comment by s3rj1k** [comment]
Could be a button in GUI that install separate SOCKS5 server application :)

**Comment by ilya-fedin** [comment]
Even more over-engineering. You propose solutions that acceptable only for unofficial clients for power users, not for an official client of a messenger for mass market.

**Comment by s3rj1k** [comment]
Well I propose a solution, you on the other hand say that this is impossible to do in any way, no offense intended.
If tdesktop team does not think that HTTPS proxy is important enough to implement, then please say so and close this request as `Wont fix`, so that public would know.

**Comment by ilya-fedin** [comment]
Yeah, it's not important enough to be implemented any time soon, but feature requests that have no plan to be implemented aren't being closed usually here.

**Comment by ilya-fedin** [comment]
I see it's closed as a duplicate. But I don't understand why Aokromes mark issues as duplicates of comments.

**Comment by s3rj1k** [comment]
I don't see any mentions of duplicate issue, it was just brutally closed :)

**Comment by ilya-fedin** [comment]
When @Aokromes postes a link to another issue or comment and closes, this means he marked the issue as a duplicate

**Comment by s3rj1k** [comment]
And probably closing them explicitly is a better way of communicating team intentions to public rather when waiting for a bot to auto-close it, in hopes that nobody cares anymore :)

**Comment by ilya-fedin** [comment]
Not really. If feature requests aren't considered to be implemented any time soon, this doesn't mean that they won't be implemented at al..

**Comment by s3rj1k** [comment]
It gives a false hope for tdesktop users, almost 3 years have past since initial report

**Comment by ilya-fedin** [comment]
probably people don't understand that feature requests are just feature requests and if no one says something will be implemented, that won't be implemented most likely any time soon

**Comment by s3rj1k** [comment]
That is because they are left open without any feedback until `бездушная машина` (bot) closes them.

There is a reason why Qt closes tickets with `Won't fix` rather then leaving them hanging there indefinitely until some auto grim-ripper closes it.

Wouldn't you say that knowing about Qt lack of interest in implementing HTTPS proxy API closes some of your open gestalts ?

**Comment by ilya-fedin** [comment]
> There is a reason why Qt closes tickets with `Won't fix` rather then leaving them hanging there indefinitely until some auto grim-ripper closes it.

Whatever you will write, this won't be changed most likely, sorry.

> Wouldn't you say that knowing about Qt lack of interest in implementing HTTPS proxy API closes some of your open gestalts ?

I can't parse that.

**Comment by s3rj1k** [comment]
So If a read your comments correctly, you say that tdesktop team does not care much about user created issues in github?

**Comment by ilya-fedin** [comment]
AFAIK there's no "tdesktop team". AFAIK john-preston is the only hired developer here and all his time goes to implementing whatever hiss bosses tell to implement him. But he is the only who reads issues here. In other words, "tdesktop team" (in form of preston) reads that, but that issue tracker is not really official and other part of Telegram doesn't see these feature requests, most likely.

**Comment by s3rj1k** [comment]
So how to make an official issue request?

Also this one makes me more sad
https://github.com/telegramdesktop/tdesktop/issues/8651#issuecomment-700412669

Seems like proxy stuff is a big tdesktop problem

**Comment by ilya-fedin** [comment]
> So how to make an official issue request?

Previously it was possible only via in-app support.
Now here is bugs.telegram.org.

**Comment by s3rj1k** [comment]
https://bugs.telegram.org/c/4358

**Comment by ilya-fedin** [comment]
(no one there closes issues as won't fix as well, btw)

**Comment by s3rj1k** [comment]
> (no one there closes issues as won't fix as well, btw)

well, lol :)

**Comment by ilya-fedin** [comment]
This issue is now a part of #25126

**Comment by Kabingeran** [comment]
Hello everyone. Please add this to the feature request too. The socks5h scheme for socks5 proxies is not supported. This causes major compatibility issues and DNS leaks. This also applies to mobile and other versions.

**Comment by ilya-fedin** [comment]
@Kabingeran wdym? telegram clients connect to the server by ip, not by hostname

**Comment by Kabingeran** [comment]
@ilya-fedin I mean, the official version of Telegram is leaking DNS. SOCKS proxies are reporting this. It is necessary to support the socks5h scheme and exclude such a scenario for security reasons.

**Comment by ilya-fedin** [comment]
> I mean, the official version of Telegram is leaking DNS. SOCKS proxies are reporting this. It is necessary to support the socks5h scheme and exclude such a scenario for security reasons.

I don't see how that could happen since Telegram doesn't use DNS to connect to the severs. I also don't understand what you mean by "support the socks5h scheme". I believe you should make a new feature request with an excessive explanation of the problem and the data which confirms the problem and allows the devs to understand where that leak happens (which domains, etc). As long as you write here, no way your case will get into work, since this request about HTTPS proxy only and it's practically wontfix due to the restrictions of the framework tdesktop uses (QtNetwork).

**Comment by Kabingeran** [comment]
@ilya-fedin Tor also reports DNS leaks. So you're now claiming that not a single domain name is in use? Then provide proof. This feature is already implemented. It just needs some improvements. I'm simply asking you to correct the text of your request and add some additional information.

**Comment by ilya-fedin** [comment]
@Kabingeran i'm not going to off-top in this thread anymore. If you want to talk about your problem, please create a new feature request. Thanks.

> I'm simply asking you to correct the text of your request and add some additional information.

This request isn't mine. I'm triager.

**Comment by s3rj1k** [comment]
Pretty sure you can vibe-code this feature in few days

**Comment by Kabingeran** [comment]
@ilya-fedin Ok. I don't use this messenger because it's not secure. So that's your problem. I hope someone makes it more secure.

**Comment by ilya-fedin** [comment]
> Pretty sure you can vibe-code this feature in few days

I tried to ask LLM by linking this issue, it found your proposition about local SOCKS proxy and said that while you're right that a prototype is reachable in a few days, it's going to be a huge privacy/vulnerability risk since a bug in such local proxy could lead to any local process being connected to it. It also said that unbound memory consumption risk is high, too.

**Comment by john-preston** [comment]
@Kabingeran Any proof? Mtproto connections to telegram don't use any domain names.

**Comment by s3rj1k** [comment]
> huge privacy/vulnerability risk

common, telegram in itself is a huge privacy risk where by default nothing is end-to-end encrypted

This feature should be treated as censorship circumvention, and not as additional layer of privacy

**Comment by ilya-fedin** [comment]
> This feature should be treated as censorship circumvention, and not as additional layer of privacy

No one talked about additional layer of privacy, the talk is about that local proxy server decreases the existing privacy.

**Comment by s3rj1k** [comment]
> No one talked about additional layer of privacy, the talk is about that local proxy server decreases the existing privacy.

How exactly it is decreasing privacy? Telegram will still tunnel encrypted MTProto no?

**Comment by ilya-fedin** [comment]
> How exactly it is decreasing privacy? Telegram will still tunnel encrypted MTProto no?

I explained this in the first answer you didn't get.

> it's going to be a huge privacy/vulnerability risk since a bug in such local proxy could lead to any local process being connected to it

**Comment by s3rj1k** [comment]
> any local process being connected to it

This is by default how this should work, any local socks proxy can be exposed to any local process running in `localhost`.

All I see excuses for implementation to happen TBH, while back in 2018 this was hard to do and maintain, now 2026 applying yet another patch to QT and have ability to build client with opt-in SOCKS and HTTPS proxy support is no longer a problem, it takes one monthly Claude Code Pro plan subscription to get all PoC patches ready.

**Comment by ilya-fedin** [comment]
> This is by default how this should work, any local socks proxy can be exposed to any local process running in `localhost`.

If you want a local proxy connectable by any local process, you don't need it embedded in tdesktop - i'm pretty sure you can use something like xray. Messenger creating a local proxy server accessible by other processes, on the other hand, is unexpected. And I bet there will be reports about that...

**Comment by s3rj1k** [comment]
> > This is by default how this should work, any local socks proxy can be exposed to any local process running in `localhost`.
> 
> If you want a local proxy connectable by any local process, you don't need it embedded in tdesktop - i'm pretty sure you can use something like xray. Messenger creating a local proxy server accessible by other processes, on the other hand, is unexpected. And I bet there will be reports about that...

I wanted native HTTPS proxy support, without running any other software, in 2026 I don't care anymore.

**Comment by Kabingeran** [comment]
@john-preston 
We did some research and concluded that most people currently use socks5 in conjunction with VLESS, VMess, Trojan or Tor. Everything else is either outdated or doesn't withstand attacks from censorship authorities. Tor also doesn't work, reporting DNS leaks. This is likely because the socks5h scheme isn't supported. The only options are either using an HTTPS proxy, which isn't supported, or using the methods I listed above.
